pkg/kubelet/volumemanager/reconciler/reconciler.go

```diff
@@ -348,74 +357,41 @@ type reconstructedVolume struct {
 	volumeSpec          *volumepkg.Spec
 	outerVolumeSpecName string
 	pod                 *v1.Pod
-	pluginIsAttachable  bool
+	attachablePlugin    volumepkg.AttachableVolumePlugin
 	volumeGidValue      string
 	devicePath          string
 	reportedInUse       bool
 	mounter             volumepkg.Mounter
 	blockVolumeMapper   volumepkg.BlockVolumeMapper
 }
 
-// reconstructFromDisk scans the volume directories under the given pod directory. If the volume is not
-// in either actual or desired state of world, or pending operation, this function will reconstruct
-// the volume spec and put it in both the actual and desired state of worlds. If no running
-// container is mounting the volume, the volume will be removed by desired state of world's populator and
-// cleaned up by the reconciler.
-func (rc *reconciler) syncStates(podsDir string) {
+// syncStates scans the volume directories under the given pod directory.
+// If the volume is not in desired state of world, this function will reconstruct
+// the volume related information and put it in both the actual and desired state of worlds.
+// For some volume plugins that cannot support reconstruction, it will clean up the existing
+// mount points since the volume is no long needed (removed from desired state)
+func (rc *reconciler) syncStates() {
 	// Get volumes information by reading the pod's directory
-	podVolumes, err := getVolumesFromPodDir(podsDir)
+	podVolumes, err := getVolumesFromPodDir(rc.kubeletPodsDir)
 	if err != nil {
 		glog.Errorf("Cannot get volumes from disk %v", err)
 		return
 	}
-
 	volumesNeedUpdate := make(map[v1.UniqueVolumeName]*reconstructedVolume)
 	for _, volume := range podVolumes {
 		reconstructedVolume, err := rc.reconstructVolume(volume)
 		if err != nil {
-			glog.Errorf("Could not construct volume information: %v", err)
+			glog.Errorf("Could not construct volume information (POD.UID %s, Volume %s): %v", volume.podName, volume.volumeSpecName, err)
+			rc.cleanupMounts(volume)
```

Comment:
The `cleanupMounts` will either fail if kubelet was restarted and a container was using the volume( `device is busy` error) or it will cause currently running container to misbehave.  Are these only options? I think gluster users have to be warned that any kubelet restart could cause the mount point to be removed.

Reply:
Alright so if pod indeed running and present in desired state, then we shouldn't have to reconstruct volume spec and the function should not have to come here.

Reply:
yes, if the pod is in desired state, we will not do anything and let the reconciler to fill the actual state.